File: rust/653.two-sum-iv-input-is-a-bst.rs
/*
 * @lc app=leetcode id=653 lang=rust
 *
 * [653] Two Sum IV - Input is a BST
 */

// @lc code=start
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn find_target(root: Option<Rc<RefCell<TreeNode>>>, k: i32) -> bool {
        use std::collections::HashSet;
        let mut set = HashSet::new();
        let mut result = false;
        fn inorder_traverse(
            node: &Option<Rc<RefCell<TreeNode>>>,
            k: &i32,
            set: &mut HashSet<i32>,
            result: &mut bool,
        ) {
            if *result {
                return;
            }
            if let Some(n) = node {
                let nb = n.borrow();
                if set.contains(&(k - nb.val)) {
                    *result = true;
                    return;
                }
                set.insert(nb.val);
                inorder_traverse(&nb.left, k, set, result);
                inorder_traverse(&nb.right, k, set, result);
            }
        }
        inorder_traverse(&root, &k, &mut set, &mut result);
        result
    }
}
// @lc code=end

File: javascript/653.two-sum-iv-input-is-a-bst.js
/*
 * @lc app=leetcode id=653 lang=javascript
 *
 * [653] Two Sum IV - Input is a BST
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {boolean}
 */
var findTarget = function (root, k) {
  const set = new Set();
  let result = false;
  function inorderTraverse(node) {
    if (result) return;
    if (node.left) inorderTraverse(node.left);
    if (set.has(k - node.val)) return (result = true);
    set.add(node.val);
    if (node.right) inorderTraverse(node.right);
  }
  inorderTraverse(root);
  return result;
};
// @lc code=end
